File: 2012-offseason/drive.lua
-- drive.lua

local linearize = require("linearize")
local math = require("math")
local pid = require("pid")
local wpilib = require("wpilib")

local pairs = pairs

local ROBOTNAME = ROBOTNAME

module(...)

local dashboard = wpilib.SmartDashboard_GetInstance()

local gyro = nil
local gyroOkay = true
local ignoreGyro = false
local gearSwitch
if ROBOTNAME == "hodgepodge" then
    gearSwitch = wpilib.Solenoid(1)
end
local brake1, brake2
brake1 = wpilib.Solenoid(4)
brake2 = wpilib.Solenoid(5)

local brakesFired = false
local isBridgeMode = false

local function limit(x)
    if x > 1 then
        return 1
    elseif x < -1 then
        return -1
    else
        return x
    end
end

local function signSquare(x)
    if x < 0 then return -x*x end
    return x*x
end

local function arcade(move, rotate)
    move = signSquare(limit(move))
    rotate = signSquare(limit(rotate))
    if move > 0 then
        if rotate > 0 then
            return move - rotate, math.max(move, rotate)
        else
            return math.max(move, -rotate), move + rotate
        end
    else
        if rotate > 0 then
            return -math.max(-move, rotate), move + rotate
        else
            return move - rotate, -math.max(-move, -rotate)
        end
    end
end

function initGyro()
    gyro = wpilib.Gyro(1, 1)
    gyro:SetSensitivity(0.00703)
    gyro:Reset()
    gyroOkay = true
    dashboard:PutBoolean("Gyro Okay", true)
end

function resetGyro()
    gyro:Reset()
    ignoreGyro = false
end

function effTheGyro()
    ignoreGyro = true
end

function getGyroAngle()
    if not gyroOkay or ignoreGyro then
        return 0
    end
    return -gyro:GetAngle()
end

function disableGyro()
    gyroOkay = false
    dashboard:PutBoolean("Gyro Okay", false)
end

local gear = "high"

local leftDriveMotor = wpilib.Victor(1)
local rightDriveMotor = wpilib.Victor(2)

function getGear()
    return gear
end

function setGear(g)
    gear = g
    if not gearSwitch then
        return
    end

    if gear == "low" then
        gearSwitch:Set(true)
    elseif gear == "high" then
        gearSwitch:Set(false)
    else
        -- Unrecognized state, default to low gear
        -- TODO: log error
        gearSwitch:Set(false)
    end
end

-- Wraps an angle (in degrees) to (-180, 180].
function normalizeAngle(theta)
    while theta > 180 do
        theta = theta - 360
    end
    while theta < -180 do
        theta = theta + 360
    end
    return theta
end

function setBrakesFired(val)
    brakesFired = val
end

local function LinearVictor(...)
    return linearize.wrap(wpilib.Victor(...))
end

function setBridgeMode(val)
    isBridgeMode = val
end


local function brakesUpdate(fire)
    brake1:Set(fire)
    brake2:Set(not fire)
end

function update(driveX,driveY)
	local leftSpeed, rightSpeed = arcade(driveY, driveX)
	leftDriveMotor:Set(-leftSpeed)
	rightDriveMotor:Set(rightSpeed)
    local distance = driveY * driveY + driveX * driveX

    local deadband = 0.1
    deadband = deadband * deadband

    brakesUpdate(brakesFired or (isBridgeMode and distance < deadband))
end

--[[
Team 254's drive code. (read as: Austin magic).

throttle : number
wheel : number
highGear : boolean
quickTurn : boolean

returns: leftSpeed, rightSpeed
--]]
local oldWheel = 0.0
local negInertiaAccumulator = 0.0
local quickStopAccumulator = 0.0
local accum
function cheesyDrive(throttle, wheel, highGear, quickTurn)
    -- constants, feel free to tune
    local turnNonlinHigh = 0.9
    local turnNonlinLow = 0.8
    local negInertiaHigh = 0.5
    local senseHigh = 1.2
    local senseLow = 1.1
    local senseCutoff = 0.1
    local negInertiaLowMore = 2.5
    local negInertiaLowLessExt = 5.0
    local negInertiaLowLess = 3.0
    local negInertiaWheelExtThreshold = 0.65
    local quickStopTimeConstant = 0.1
    local quickStopLinearPowerThreshold = 0.2
    local quickStopStickScalar = 5.0
    -- /constants

    -- shortcuts
    local sin = math.sin
    local pi_2 = math.pi / 2.0
    local fabs = math.abs
    -- /shortcuts

    local wheelNonLinearity
    local negInertia = wheel - oldWheel
    oldWheel = wheel

    if highGear then
        wheelNonLinearity = turnNonlinHigh
        -- Apply a sin function that's scaled to make it feel better
        wheel = sin(pi_2 * wheelNonLinearity * wheel) / sin(pi_2 * wheelNonLinearity)
        wheel = sin(pi_2 * wheelNonLinearity * wheel) / sin(pi_2 * wheelNonLinearity)
    else
        wheelNonLinearity = turnNonlinLow
        -- Apply a sin function that's scaled to make it feel better
        wheel = sin(pi_2 * wheelNonLinearity * wheel) / sin(pi_2 * wheelNonLinearity)
        wheel = sin(pi_2 * wheelNonLinearity * wheel) / sin(pi_2 * wheelNonLinearity)
        wheel = sin(pi_2 * wheelNonLinearity * wheel) / sin(pi_2 * wheelNonLinearity)
    end

    local sensitivity
    local negInertiaScalar
    if highGear then
        negInertiaScalar = negInertiaHigh
        sensitivity = senseHigh
    else
        if wheel * negInertia > 0 then
            negInertiaScalar = negInertiaLowMore
        elseif fabs(wheel) > negInertiaWheelExtThreshold then
            negInertiaScalar = negInertiaLowLessExt
        else
            negInertiaScalar = negInertiaLowLess
        end
        sensitivity = senseLow
        if fabs(throttle) > senseCutoff then
            sensitivity = 1 - (1 - sensitivity) / fabs(throttle)
        end
    end

    local negInertiaPower = negInertia * negInertiaScalar
    negInertiaAccumulator = negInertiaAccumulator + negInertiaPower

    wheel = wheel + negInertiaAccumulator
    negInertiaAccumulator = accum(negInertiaAccumulator)

    local linearPower = throttle

    -- Quickturn!
    local overPower
    local angularPower
    if quickTurn then
        if fabs(linearPower) < quickStopLinearPowerThreshold then
            quickStopAccumulator = (1 - quickStopTimeConstant) * quickStopAccumulator + quickStopTimeConstant * limit(wheel) * quickStopStickScalar
        end
        overPower = 1.0
        sensitivity = 1.0
        angularPower = wheel
    else
        overPower = 0.0
        angularPower = fabs(throttle) * wheel * sensitivity - quickStopAccumulator
        quickStopAccumulator = accum(quickStopAccumulator)
    end

    -- Compute left/right speeds
    local leftPwm = linearPower + angularPower
    local rightPwm = linearPower - angularPower

    if leftPwm > 1.0 then
        rightPwm = rightPwm - (overPower * (leftPwm - 1))
        leftPwm = 1.0
    elseif rightPwm > 1.0 then
        leftPwm = leftPwm - (overPower * (rightPwm - 1))
        rightPwm = 1.0
    elseif leftPwm < -1.0 then
        rightPwm = rightPwm + (overPower * (-1 - leftPwm))
        leftPwm = -1.0
    elseif rightPwm < -1.0 then
        leftPwm = leftPwm + (overPower * (-1 - rightPwm))
        rightPwm = -1.0
    end
    return leftPwm, rightPwm
end
function accum(val)
    if val > 1 then
        return val - 1
    elseif val < -1 then
        return val + 1
    else
        return 0
    end
end

--[[
drive victors:
left 2,
right 1,
Intake victors:
3
shooter:
4
Cheater:
5
Vertical:
6
SENSORS:
pressure switch:
2
banner sensor:
3
Pneumatics:
drive:
1
intake:
2
stinger:
3
--]]
-- vim: ft=lua et ts=4 sts=4 sw=4

File: 2013-cpp/SimpleTemplate/drive.cpp
#include "WPILib.h"
#include "drive.hpp"
#include <math.h>
#include <algorithm>

Drive::Drive()
{
    frontLeftDrive = new Talon(1);
    frontRightDrive = new Talon(2);
    backLeftDrive = new Talon(9);
    backRightDrive = new Talon(10);

    lowGear = new Solenoid(3);
    kickUp = new Solenoid(2);
    oldWheel = 0.0;
    negInertiaAccumulator = 0.0;
    quickStopAccumulator = 0.0;
    isLowGear = false;
    isKickUp = false;
}

void Drive::setKickUp(bool k)
{
    isKickUp = k;
}

// Divide by for 2.778 scaling
void Drive::setBackWheelsDown(bool d)
{
    isLowGear = d;
}

void Drive::setHighGear(bool g)
{
    if ((isLowGear) || (isKickUp))
    {
    isLowGear = isLowGear;
    isKickUp = isKickUp;
    }
    else
    {
        isLowGear = g;
        isKickUp = g;
    }
}

void Drive::setDriveMotors(float left, float right)
{
    if ((isKickUp) && (! isLowGear))
    {
        frontLeftDrive->Set(-limit(left));
        backLeftDrive->Set(-limit(left));

        frontRightDrive->Set(limit(right / 2.778));
        backRightDrive->Set(limit(right / 2.778));
    }
    else if ((! isKickUp) && (isLowGear))
    {
        frontLeftDrive->Set(-limit(left / 2.778));
        backLeftDrive->Set(-limit(left / 2.778));

        frontRightDrive->Set(limit(right));
        backRightDrive->Set(limit(right));
    }
    else
    {
        frontLeftDrive->Set(-limit(left));
        backLeftDrive->Set(-limit(left));

        frontRightDrive->Set(limit(right));
        backRightDrive->Set(limit(right));
    }

}

float Drive::limit(float x)
{
    if (x > 1)
        return 1;
    else if (x < -1)
        return -1;
    else
        return x;
}

void Drive::CheesyDrive(double throttle, double wheel, bool highGear, bool quickTurn) {
  bool isQuickTurn = quickTurn;
  float turnNonlinHigh = 0.9;
  float turnNonlinLow = 0.8;
  float negInertiaHigh = 0.5;
  float senseHigh = 1.2;
  float senseLow = 0.6;
  float senseCutoff = 0.1;
  float negInertiaLowMore = 2.5;
  float negInertiaLowLessExt = 5.0;
  float negInertiaLowLess = 3.0;
  float quickStopTimeConstant = 0.1;
  float quickStopLinearPowerThreshold = 0.2;
  float quickStopStickScalar = 5.0;

  double wheelNonLinearity;

  double negInertia = wheel - oldWheel;
  oldWheel = wheel;

  double M_PI = 3.141592;

  if (highGear) {
    wheelNonLinearity = turnNonlinHigh;
    // Apply a sin function that's scaled to make it feel better.
    wheel = sin(M_PI / 2.0 * wheelNonLinearity * wheel) / sin(M_PI / 2.0 * wheelNonLinearity);
    wheel = sin(M_PI / 2.0 * wheelNonLinearity * wheel) / sin(M_PI / 2.0 * wheelNonLinearity);
  } else {
    wheelNonLinearity = turnNonlinLow;
    // Apply a sin function that's scaled to make it feel better.
    wheel = sin(M_PI / 2.0 * wheelNonLinearity * wheel) / sin(M_PI / 2.0 * wheelNonLinearity);
    wheel = sin(M_PI / 2.0 * wheelNonLinearity * wheel) / sin(M_PI / 2.0 * wheelNonLinearity);
    wheel = sin(M_PI / 2.0 * wheelNonLinearity * wheel) / sin(M_PI / 2.0 * wheelNonLinearity);
  }

  double leftPwm, rightPwm, overPower;
  float sensitivity = 1.7;

  float angularPower;
  float linearPower;

  // Negative inertia!
  static double negInertiaAccumulator = 0.0;
  double negInertiaScalar;
  if (highGear) {
    negInertiaScalar = negInertiaHigh;
    sensitivity = senseHigh;
  } else {
    if (wheel * negInertia > 0) {
      negInertiaScalar = negInertiaLowMore;
    } else {
      if (fabs(wheel) > 0.65) {
        negInertiaScalar = negInertiaLowLessExt;
      } else {
        negInertiaScalar = negInertiaLowLess;
      }
    }
    sensitivity = senseLow;

    if (fabs(throttle) > senseCutoff) {
      sensitivity = 1 - (1 - sensitivity) / fabs(throttle);
    }
  }
  double negInertiaPower = negInertia * negInertiaScalar;
  negInertiaAccumulator += negInertiaPower;

  wheel = wheel + negInertiaAccumulator;
  if(negInertiaAccumulator > 1)
    negInertiaAccumulator -= 1;
  else if (negInertiaAccumulator < -1)
    negInertiaAccumulator += 1;
  else
    negInertiaAccumulator = 0;

  linearPower = throttle;

  // Quickturn!
  if (isQuickTurn) {
    if (fabs(linearPower) < quickStopLinearPowerThreshold) {
      double alpha = quickStopTimeConstant;
      quickStopAccumulator = (1 - alpha) * quickStopAccumulator + alpha * limit(wheel) * quickStopStickScalar;
    }
    overPower = 1.0;
    if (highGear) {
      sensitivity = 1.0;
    } else {
      sensitivity = 1.0;
    }
    angularPower = wheel;
  } else {
    overPower = 0.0;
    angularPower = fabs(throttle) * wheel * sensitivity - quickStopAccumulator;
    if (quickStopAccumulator > 1) {
      quickStopAccumulator -= 1;
    } else if (quickStopAccumulator < -1) {
      quickStopAccumulator += 1;
    } else {
      quickStopAccumulator = 0.0;
    }
  }

  rightPwm = leftPwm = linearPower;
  leftPwm += angularPower;
  rightPwm -= angularPower;

  if (leftPwm > 1.0) {
    rightPwm -= overPower * (leftPwm - 1.0);
    leftPwm = 1.0;
  } else if (rightPwm > 1.0) {
    leftPwm -= overPower * (rightPwm - 1.0);
    rightPwm = 1.0;
  } else if (leftPwm < -1.0) {
    rightPwm += overPower * (-1.0 - leftPwm);
    leftPwm = -1.0;
  } else if (rightPwm < -1.0) {
    leftPwm += overPower * (-1.0 - rightPwm);
    rightPwm = -1.0;
  }

  setDriveMotors(leftPwm, rightPwm);
  setHighGear(highGear);
}

void Drive::mecanumDrive(double x, double y, double z)
{
    double theta = 0.0;
    double temp = y*cos(theta) - x*sin(theta);
    x = y*sin(theta) + x*cos(theta);
    y = temp;

    int numWheels = 3;
    float wheels[4];
    wheels[0] = y + x + z;
    wheels[1] = y - x - z;
    wheels[2] = y - x + z;
    wheels[3] = y + x - z;

    float max = 0.0;
    for (int i=0; i <= numWheels; i++)
    {
        max = fabs(wheels[i]);
        for (int j=0; j <= numWheels; j++)
        {
            if (fabs(wheels[j]) > max)
            {
                max = fabs(wheels[j]);
            }
        }
    }

    if (max > 1)
    {
        wheels[0] /= max;
        wheels[1] /= max;
        wheels[2] /= max;
        wheels[3] /= max;
    }

    frontLeftDrive->Set(wheels[0]);
    frontRightDrive->Set(wheels[1]);
    backLeftDrive->Set(wheels[2]);
    backRightDrive->Set(wheels[3]);
}

void Drive::update(double DriveX, double DriveY, double DriveZ, bool Gear, bool quickTurn)
{
    if ((isLowGear) && (isKickUp))
    {
        CheesyDrive(DriveY, (DriveX * 0), Gear, false);
    }
    else
    {
        CheesyDrive(DriveY, DriveX, Gear, quickTurn);
    }

    lowGear->Set(isLowGear);
    kickUp->Set(isKickUp);
}

